### PR TITLE
Rate limiter, pagination helper, improved API error reporting

### PR DESCRIPTION
**Pull request by adrusi**

### CODE REVIEW

**Comment by adrusi** [comment]
I can't seem to reproduce these travis errors locally, not on a freshly cloned repository. Any ideas?

File: test/test-transloadit-client.coffee [hunk 1]
```diff
@@ -4,31 +4,6 @@ expect            = require("chai").expect
 TransloaditClient = require "../src/TransloaditClient"
 
 describe "TransloaditClient", ->
-  describe "constructor", ->
```
Comment (tim-kos) [inline test/test-transloadit-client.coffee]: Why did you remove these tests?

File: test/test-transloadit-client.coffee [hunk 2]
```diff
@@ -59,304 +34,6 @@ describe "TransloaditClient", ->
...
-        cb ERR
-
-        gently.expect client, "_remoteJson", (opts2, cb2) ->
-          expect(opts2.url).to.eql RESULT.assembly_url
-
-        cb null, RESULT
-
-      client.getAssembly ASSEMBLY_ID, CB
-      expect(errCalls).to.equal 1
-      expect(calls).to.equal 1
```
Comment (tim-kos) [inline test/test-transloadit-client.coffee]: Why did you remove all these tests?

File: src/TransloaditClient.coffee [hunk 3]
```diff
@@ -21,6 +25,7 @@ class TransloaditClient
     @_authSecret = opts.authSecret
     @_service    = opts.service || "api2.transloadit.com"
     @_region     = opts.region || "us-east-1"
+    @_useSsl     = opts.useSsl
```
Comment (tim-kos) [inline src/TransloaditClient.coffee]: Is @_useSsl used anywhere? Or is this just for convenience for the future?

File: src/PaginationStream.coffee [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+      if err?
+        return @emit "error", err
+
+      @_nitems = result.count
+
+      @_items = (result.items[i] for i in [result.items.length-1 .. 0])
+
+      return @_read()
+    
+module.exports = PaginationStream
```
Comment (tim-kos) [inline src/PaginationStream.coffee]: I like this! :)

File: README.md [hunk 5]
```diff
@@ -94,6 +94,11 @@ Retrieves an array of assemblies according to the given `params`.
 
 Valid params can be page, pagesize, type, fromdate and todate. Please consult the [Transloadit API docs](https://transloadit.com/docs/api-docs/#retrieve-assembly-list) for details.
 
+#### TransloaditClient.streamAssemblies(params)
+
+Creates an objectMode readable stream that automates handling of listAssembly
+pagination.
```
Comment (tim-kos) [inline README.md]: Let's add short code samples here (basically the test that you added).

File: src/TransloaditClient.coffee [hunk 6]
```diff
@@ -57,37 +62,21 @@ class TransloaditClient
...
-        if err?
-          return cb err
-        
-        if !result.assembly_url?
-          if operation.retry new Error "failed to retrieve assembly_url"
-            return
-          
-          return cb operation.mainError()
-
-        finallyDelete result.assembly_url
```
Comment (tim-kos) [inline src/TransloaditClient.coffee]: 👍

File: src/TransloaditClient.coffee [hunk 7]
```diff
@@ -129,21 +118,29 @@ class TransloaditClient
...
+
   getAssembly: (assemblyId, cb) ->
     opts =
       url: @_serviceUrl() + "/assemblies/#{assemblyId}"
 
-    @_remoteJson opts, (err, result) =>
-      if err
-        return cb err
+    operation = retry.operation retries: 5, factor: 3.28,
+      minTimeout: 1 * 1000, maxTimeout: 8 * 1000
```
Comment (tim-kos) [inline src/TransloaditClient.coffee]: Let's use an opts object here or keep the options one on each line for readability please.

File: src/TransloaditClient.coffee [hunk 8]
```diff
@@ -275,15 +272,20 @@ class TransloaditClient
...
       maxTimeout : 8 * 1000
-    )
 
     operation.attempt =>
       @__remoteJson opts, (err, result) ->
+        if err? && err.error == "RATE_LIMIT_REACHED"
+          console.warn "Rate limit reached, retrying request in #{err.info.retryIn} seconds."
+          # FIXME uses private internals of node-retry
+          operation._timeouts.unshift 1000 * err.info.retryIn
+          return operation.retry err
```
Comment (tim-kos) [inline src/TransloaditClient.coffee]: Hm, this is a tough one. On the one hand you could use retry with its fairly new `forever` option. I would then add a `stop()` method to it.

The downside of this would be that we'd only have a fixed timeout / set of timeouts, and could not put `err.info.retryIn` to good use. Maybe I should add an `addTimeout()` function to node-retry? What do you think?

File: test/integration.coffee [hunk 9]
```diff
@@ -82,6 +82,36 @@ describe "API integration", ->
...
+          client.getAssembly result.assembly_id, (err, result) ->
+            if err? || result.error?
+              return reproduce nattempts - 1
+
+            expect(result).to.have.property("assembly_url").that.exist
+            reproduce nattempts - 1
+
+      # attempt to reproduce the incomplete status response 100 times
+      for x in [1..nbranches]
+        reproduce 100 / nbranches
```
Comment (tim-kos) [inline test/integration.coffee]: 👌

File: test/test-pagination-stream.coffee [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+      process.nextTick ->
+        cb null, pages[pageno - 1]
+
+    stream.pipe toArray (array) ->
+      expected = _.flatten (page.items for page in pages), true
+
+      expect(array).to.deep.equal expected
+      done()
+
+    stream.resume()
```
Comment (tim-kos) [inline test/test-pagination-stream.coffee]: Much 😍 for this.
